### PR TITLE
Fix signature polymorphism check

### LINKED ISSUE #58: TACAI fails on OpenJDK 15
On OpenJDK 15, the PurityTests fail with a NullPointerException during TACAI, which didn't happen in OpenJDK 11.

### PR DESCRIPTION
**Pull request by errt**
Fixes #58

### CODE REVIEW

File: OPAL/br/src/main/scala/org/opalj/br/analyses/ProjectLike.scala [hunk 1]
```diff
@@ -606,11 +606,10 @@ abstract class ProjectLike extends ClassFileRepository { project ⇒
...
     //TODO add method that lookup the defining class type
     def isSignaturePolymorphic(definingClassType: ObjectType, method: Method): Boolean = {
-        (
-            (definingClassType eq ObjectType.MethodHandle) ||
-            (definingClassType eq ObjectType.VarHandle)
-        ) &&
-            method.isNativeAndVarargs &&
+        (definingClassType eq ObjectType.MethodHandle) &&
+            method.descriptor == SignaturePolymorphicMethodObject &&
+            (method.name == "invoke" || method.name == "invokeExact") ||
```
Comment (mreif) [inline OPAL/br/src/main/scala/org/opalj/br/analyses/ProjectLike.scala]: Have you checked that one? Last time I checked (not today) the operations were not performed, when invokeExact is called - so it should not be signature polymorphic

File: OPAL/br/src/main/scala/org/opalj/br/analyses/ProjectLike.scala [hunk 2]
```diff
@@ -285,7 +285,7 @@ abstract class ProjectLike extends ClassFileRepository { project ⇒
                 // we have to avoid endless recursion if we can't find the target method
                 if ((MethodHandleSubtypes.contains(receiverType) ||
                     VarHandleSubtypes.contains(receiverType)) &&
-                    !isSignaturePolymorphic(receiverType, name, descriptor)) {
+                    !isSignaturePolymorphic(receiverType, descriptor)) {
                     // At least in Java 15 the signature polymorphic methods are not overloaded and
                     // it actually doesn't make sense to do so. Therefore we decided to only
                     // make this lookup if strictly required.
```
Comment (mreif) [inline OPAL/br/src/main/scala/org/opalj/br/analyses/ProjectLike.scala:291]: sounds reasonable

**Review by mreif — COMMENTED** [review]
Could you double check the "invokeExact" behavior please?

File: OPAL/br/src/main/scala/org/opalj/br/analyses/ProjectLike.scala (hunk 1 above)
Comment (errt) [reply OPAL/br/src/main/scala/org/opalj/br/analyses/ProjectLike.scala]: You can check https://docs.oracle.com/en/java/javase/15/docs/api/java.base/java/lang/invoke/MethodHandle.html to see that invoke and invokeExact are signature polymorphic. You may be thinking about the wrong concept here. Signature polymorphism just means that invokevirtual does not check the actual signature. Adaptation of parameters is performed by invoke (and non-signature-polymorphic invokeWithArguments!) only, but this is not connected to signature polymorphism. However, I checked the JVM spec again and will remove this name check here again as it is not necessary - here we have the method available and can check the native and varargs flags and that exactly matches the definition of the JVM spec. The name check is necessary for the other method however where only the signature is known.

**Review by mreif — APPROVED** [review]

**Pull request merged**